client/compliance_client.go

```diff
@@ -585,6 +654,894 @@ func (c *complianceGRPCClient) CancelOperation(ctx context.Context, req *longrun
+
+	if httpRsp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(httpRsp.Status)
+	}
+
+	buf, err := ioutil.ReadAll(httpRsp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	unm := protojson.UnmarshalOptions{AllowPartial: true, DiscardUnknown: true}
+	rsp := &genprotopb.RepeatResponse{}
+
+	return rsp, unm.Unmarshal(buf, rsp)
+}
+
+// RepeatDataQuery this method echoes the ComplianceData request. This method exercises
+// sending all request fields as query parameters.
+func (c *complianceRESTClient) RepeatDataQuery(ctx context.Context, req *genprotopb.RepeatRequest, opts ...gax.CallOption) (*genprotopb.RepeatResponse, error) {
+	m := protojson.MarshalOptions{AllowPartial: true, EmitUnpopulated: true}
+	jsonReq, err := m.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	baseUrl, _ := url.Parse(c.endpoint)
+	baseUrl.Path += fmt.Sprintf("/v1beta1/repeat:query")
+
+	params := url.Values{}
+	if req.GetInfo().GetFBool() {
+		params.Add("info.fBool", fmt.Sprintf("%v", req.GetInfo().GetFBool()))
+	}
+	if req.GetInfo().GetFBytes() != nil {
+		params.Add("info.fBytes", fmt.Sprintf("%v", req.GetInfo().GetFBytes()))
+	}
+	if req.GetInfo().GetFChild().GetFBool() {
+		params.Add("info.fChild.fBool", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFBool()))
+	}
+	if req.GetInfo().GetFChild().GetFChild().GetFBool() {
+		params.Add("info.fChild.fChild.fBool", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFChild().GetFBool()))
+	}
+	if req.GetInfo().GetFChild().GetFChild().GetFDouble() != 0 {
+		params.Add("info.fChild.fChild.fDouble", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFChild().GetFDouble()))
+	}
+	if req.GetInfo().GetFChild().GetFChild().GetFString() != "" {
+		params.Add("info.fChild.fChild.fString", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFChild().GetFString()))
+	}
+	if req.GetInfo().GetFChild().GetFContinent() != 0 {
+		params.Add("info.fChild.fContinent", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFContinent()))
+	}
+	if req.GetInfo().GetFChild().GetFDouble() != 0 {
+		params.Add("info.fChild.fDouble", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFDouble()))
+	}
+	if req.GetInfo().GetFChild().GetFFloat() != 0 {
+		params.Add("info.fChild.fFloat", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFFloat()))
+	}
+	if req.GetInfo().GetFChild().GetFString() != "" {
+		params.Add("info.fChild.fString", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetFString()))
+	}
+	if req.GetInfo().GetFChild().GetPBool() {
+		params.Add("info.fChild.pBool", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetPBool()))
+	}
+	if req.GetInfo().GetFChild().GetPChild().GetFBool() {
+		params.Add("info.fChild.pChild.fBool", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetPChild().GetFBool()))
+	}
+	if req.GetInfo().GetFChild().GetPChild().GetFDouble() != 0 {
+		params.Add("info.fChild.pChild.fDouble", fmt.Sprintf("%v", req.GetInfo().GetFChild().GetPChild().GetFDouble()))
+	}
+	if req.GetInfo().GetFChild().GetPChild().GetFString() != "" {
```

Comment:
Note that if `PChild` (which is `optional`) has not been set at all, we don't need to dive into the fields of  the empty message returned by by `GetPChild()`. And also, this could cause problems or unwieldiness for recursive definitions. See my other comments.

Reply:
Damn, you're right. I completely neglected that possibility.

Reply:
The more I think about this the more I'm in favor of not handling this case specially. It looks like go does not provide any mechanism for determining whether an optional field has been set.

Consider the following:
```proto
syntax = "proto3";
package test.thing;

option go_package = "server/genproto";

message Inner {}

message Outer {
        Inner required_inner = 1;
        optional Inner optional_inner = 2;
        int32 required_int = 3;
        optional int32 optional_int = 4;
        oneof variant {
              Inner inner_variant = 5;
              int32 int32_variant = 6;
        }
}
```

This gives us the following:
```go
type Outer struct {
// ...
	RequiredInner *Inner `protobuf:"bytes,1,opt,name=required_inner,json=requiredInner,proto3" json:"required_inner,omitempty"`
	OptionalInner *Inner `protobuf:"bytes,2,opt,name=optional_inner,json=optionalInner,proto3,oneof" json:"optional_inner,omitempty"`
// ...
}

func (x *Outer) GetRequiredInner() *Inner {
	if x != nil {
		return x.RequiredInner
	}
	return nil
}

func (x *Outer) GetOptionalInner() *Inner {
	if x != nil {
		return x.OptionalInner
	}
	return nil
}
```

That's all the support we get for `optional_inner`. There does not seem to be a `HasOptionalInner` method.

Recursive messages are already handled by logic in `getLeafs`, so I don't think that's a huge problem.

Reply:
This is one of those moments where it is OK to directly reference the field. `if x.OptionalInner != nil`.

Setting & checking presence is an acceptable use of direct reference. Retrieving a value should use the accessor, as best practice.